go/vt/vtgate/scatter_conn.go

```diff
@@ -299,17 +285,19 @@ func (stc *ScatterConn) ExecuteBatch(
 
 			var err error
 			startTime, statsKey := stc.startAction("ExecuteBatch", target)
-			defer stc.endAction(startTime, allErrors, statsKey, &err)
+			defer stc.endAction(startTime, allErrors, statsKey, &err, session)
 
 			shouldBegin, transactionID := transactionInfo(target, session, false)
 			var innerqrs []sqltypes.Result
 			if shouldBegin {
 				innerqrs, transactionID, err = stc.gateway.BeginExecuteBatch(ctx, target, req.Queries, asTransaction, options)
 				if transactionID != 0 {
-					session.Append(&vtgatepb.Session_ShardSession{
+					if appendErr := session.Append(&vtgatepb.Session_ShardSession{
 						Target:        target,
 						TransactionId: transactionID,
-					})
+					}); appendErr != nil {
+						err = appendErr
```

Comment:
This would return the error of session.Append instead of the error of gateway.BeginExecuteBatch. In the case where both are not nil, we should probably return the other one?